pkg/clients/dns/dns.go

```diff
@@ -0,0 +1,309 @@
+	azure "github.com/crossplane/provider-azure/pkg/clients"
+)
+
+const (
+	// RecordSetSuccessfulState represents a RecordSet that is ready and has a successful last operation.
+	RecordSetSuccessfulState = "Succeeded"
+	// RecordSetDeletingState represents a RecordSet that is deleting.
+	RecordSetDeletingState = "Deleting"
+)
+
+// ZoneAPI represents the API interface for a DNS Zone client
+type ZoneAPI interface {
+	Get(ctx context.Context, z *v1alpha1.Zone) (dns.Zone, error)
+	CreateOrUpdate(ctx context.Context, z *v1alpha1.Zone) error
+	Delete(ctx context.Context, z *v1alpha1.Zone) error
+}
+
+// ZoneClient is the concrete implementation of the ZoneAP interface for DNS Zone that calls Azure API.
+type ZoneClient struct {
+	dns.ZonesClient
+}
+
+// NewZoneClient creates and initializes a ZoneClient instance.
+func NewZoneClient(cl dns.ZonesClient) *ZoneClient {
+	return &ZoneClient{
+		ZonesClient: cl,
+	}
+}
+
+// Get retrieves the requested DNS Zone
+func (c *ZoneClient) Get(ctx context.Context, z *v1alpha1.Zone) (dns.Zone, error) {
+	return c.ZonesClient.Get(ctx, z.Spec.ForProvider.ResourceGroupName, meta.GetExternalName(z))
+}
+
+// CreateOrUpdate creates or updates a DNS Zone
+func (c *ZoneClient) CreateOrUpdate(ctx context.Context, z *v1alpha1.Zone) error {
+	_, err := c.ZonesClient.CreateOrUpdate(ctx, z.Spec.ForProvider.ResourceGroupName, meta.GetExternalName(z),
+		NewZoneParameters(z), "", "")
+
+	return err
+}
+
+// Delete deletes the given DNS Zone
+func (c *ZoneClient) Delete(ctx context.Context, z *v1alpha1.Zone) error {
+	_, err := c.ZonesClient.Delete(ctx, z.Spec.ForProvider.ResourceGroupName, meta.GetExternalName(z), z.Status.AtProvider.Etag)
+
+	return err
+}
+
+// UpdateZoneStatusFromAzure updates the status related to the external
+// Azure DNS Zone in the ZoneStatus
+func UpdateZoneStatusFromAzure(v *v1alpha1.Zone, az dns.Zone) {
+	v.Status.AtProvider.ID = azure.ToString(az.ID)
+	v.Status.AtProvider.Etag = azure.ToString(az.Etag)
+	v.Status.AtProvider.Name = azure.ToString(az.Name)
+	v.Status.AtProvider.Type = azure.ToString(az.Type)
+	v.Status.AtProvider.MaxNumberOfRecordSets = azure.Int64ToInt(az.MaxNumberOfRecordSets)
+	v.Status.AtProvider.NameServers = azure.ToStringArray(az.NameServers)
+	v.Status.AtProvider.NumberOfRecordSets = azure.Int64ToInt(az.NumberOfRecordSets)
+}
+
+// NewZoneParameters returns an Azure DNS Zone object.
+func NewZoneParameters(r *v1alpha1.Zone) dns.Zone {
+	res := dns.Zone{
+		Name: azure.ToStringPtr(meta.GetExternalName(r)),
+		ZoneProperties: &dns.ZoneProperties{
+			ZoneType: dns.ZoneType(r.Spec.ForProvider.ZoneType),
+		},
+	}
+
+	if r.Spec.ForProvider.RegistrationVirtualNetworks != nil {
+		res.RegistrationVirtualNetworks = newSubResourceParameters(r.Spec.ForProvider.RegistrationVirtualNetworks)
+	}
+
+	if r.Spec.ForProvider.ResolutionVirtualNetworks != nil {
+		res.ResolutionVirtualNetworks = newSubResourceParameters(r.Spec.ForProvider.ResolutionVirtualNetworks)
```

Comment:
nit: `nil` check could be moved to `newSubResourceParameters`, i.e., inside that function if the passed param is `nil`, `return nil`.